### ISSUE TITLE
Make sure range training works with custom classes

### ISSUE THREAD

**Issue by thomasp85** [r-lib/scales#480]
Only require `range()` and `as.numeric()` to be supported by the custom vector class

**Comment by thomasp85** [comment]
For discrete custom vectors require a `levels()` method

**Comment by fabian-s** [comment]
@teunbrand  @thomasp85 

https://github.com/r-lib/scales/blob/04fc3331af14da0d5ca1919f37c6c35d6fc512bb/R/scale-continuous.R#L54


the check seems overly broad and unspecific to me if the goal of the change discussed here was to actually "make sure range training works with custom classes", many of which will return `"list"` from `typeof` 

Is this really the desired behavior? 

Case in point: any `vctrs`-class that of `rcrd` type will trigger this error, even if they implement suitable numeric/continous `range`-methods -- see also our previous discussion in  https://github.com/tidyverse/ggplot2/issues/6184.

can this be changed to something like

`typeof(new) %in% c("logical", "character")` 

or 

`!typeof(unlist(new)) %in% c("integer", "double")`

e.g.?  that would seem to avoid these false positives and still reliably catch "things that should have been a `factor`".

**Comment by teunbrand** [comment]
It needs a `range()` method that returns a numeric result. For example:

``` r
rcd <- vctrs::new_rcrd(list(a = c(0, 5, 10), b = c("A", "B", "C")), class = "foo")

range.foo <- function(x, ...) range(vctrs::field(x, "a"), ...)

rng <- scales::ContinuousRange$new()
rng$train(rcd)
rng$range
#> [1]  0 10
```

<sup>Created on 2025-04-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by fabian-s** [comment]
thanks for your reply. this has helped me figure out an issue with my implementation of the `range`-method, which didn't include a `finite` argument so `...`-matching led to weird errors (that got suppressed).

but even after fixing that, it still remains the case that scale-training for `ggplot2`-layers on `vctrs_vctr` objects that also inherit from `list`  does not work, because the `train` method for (e.g.) `ScaleContinuous`  checks the `typeof` the original object, instead of the type of its `range`:

https://github.com/tidyverse/ggplot2/blob/3aec34a97b4567b5daf5f958870a9efeba3a4164/R/scale-.R#L683:L700

this makes flexible, robust custom geoms etc for vctrs-objects very hard to achieve.
is there a plan to migrate the more robust code now in `scales` to these `ggproto`-classes as well?

**Comment by teunbrand** [comment]
Good point about the ggplot2 scales, I had forgotten these needed to be adapted too. Let's reopen the original issue for that part.
